### PR TITLE
log errors in decoder

### PR DESCRIPTION
**Pull request by graebm**
- log errors in the decoder
- misc modifications while I'm in there and thinking about it

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: source/decode.c [hunk 1]
```diff
@@ -497,17 +616,22 @@ static int s_linestate_response(struct aws_http_decoder *decoder, struct aws_byt
...
     if (!aws_byte_cursor_eq(&version, &version_expected)) {
+        AWS_LOGF_ERROR(
+            AWS_LS_HTTP_STREAM, "id=%p: Incoming response uses unsupported HTTP version.", decoder->logging_id);
         return aws_raise_error(AWS_ERROR_HTTP_PARSE);
     }
 
-    int64_t code_val;
-    int err = s_read_int64(code, &code_val);
+    size_t code_val;
+    err = s_read_size(code, &code_val, 10);
```
Comment (JonathanHenson) [inline source/decode.c]: why 10?

File: source/decode.c [hunk 2]
```diff
@@ -404,48 +467,89 @@ static int s_linestate_header(struct aws_http_decoder *decoder, struct aws_byte_
...
+                } else if (
+                    aws_string_eq_byte_cursor_ignore_case(s_transfer_coding_gzip, &coding) ||
+                    aws_string_eq_byte_cursor_ignore_case(s_transfer_coding_x_gzip, &coding)) {
+                    /* A recipient SHOULD consider "x-gzip" to be equivalent to "gzip". RFC-7230 4.2.3 */
+                    decoder->transfer_encoding |= AWS_HTTP_TRANSFER_ENCODING_GZIP;
+
+                } else if (coding.len > 0) {
+                    AWS_LOGF_ERROR(
+                        AWS_LS_HTTP_STREAM,
+                        "id=%p: Incoming transfer-encoding header lists unrecognized coding.",
```
Comment (justinboswell) [inline source/decode.c]: [nitpick] Wouldn't it be an unrecognied encoding?

File: source/decode.c [hunk 3]
```diff
@@ -463,15 +567,27 @@ static int s_linestate_header(struct aws_http_decoder *decoder, struct aws_byte_
...
+    }
+
     struct aws_byte_cursor method = cursors[0];
     struct aws_byte_cursor uri = cursors[1];
     struct aws_byte_cursor version = cursors[2];
 
     struct aws_byte_cursor version_expected = aws_http_version_to_str(AWS_HTTP_VERSION_1_1);
     if (!aws_byte_cursor_eq(&version, &version_expected)) {
+        AWS_LOGF_ERROR(
+            AWS_LS_HTTP_STREAM, "id=%p: Incoming request uses unsupported HTTP version.", decoder->logging_id);
```
Comment (justinboswell) [inline source/decode.c]: Why not log the unexpected version?

File: source/decode.c [hunk 4]
```diff
@@ -579,3 +703,7 @@ int aws_http_decoder_get_encoding_flags(const struct aws_http_decoder *decoder)
 size_t aws_http_decoder_get_content_length(const struct aws_http_decoder *decoder) {
     return decoder->content_length;
 }
+
+void aws_http_decoder_set_logging_id(struct aws_http_decoder *decoder, void *id) {
```
Comment (justinboswell) [inline source/decode.c]: AFAICT this function is never called. Is that intentional?

File: source/decode.c [hunk 5]
```diff
@@ -497,17 +616,22 @@ static int s_linestate_response(struct aws_http_decoder *decoder, struct aws_byt
...
     if (err) {
+        AWS_LOGF_ERROR(
+            AWS_LS_HTTP_STREAM, "id=%p: Failed to parse status code from incoming response.", decoder->logging_id);
         return AWS_OP_ERR;
     }
 
     /* Status-code is a 3-digit integer. RFC7230 section 3.1.2 */
-    if (code.len != 3 || code_val < 0 || code_val > 999) {
+    if (code.len != 3 || code_val > 999) {
+        AWS_LOGF_ERROR(AWS_LS_HTTP_STREAM, "id=%p: Incoming response has invalid status code", decoder->logging_id);
```
Comment (justinboswell) [inline source/decode.c]: Worth logging some form of what that code was?

File: source/decode.c [hunk 6]
```diff
@@ -404,48 +467,89 @@ static int s_linestate_header(struct aws_http_decoder *decoder, struct aws_byte_
...
+                    AWS_LS_HTTP_STREAM,
+                    "id=%p: Incoming headers for both content-length and transfer-encoding received. This is illegal.",
+                    decoder->logging_id);
+                return aws_raise_error(AWS_ERROR_HTTP_PARSE);
+            }
+
+            if (s_read_size(header.value_data, &decoder->content_length, 10) != AWS_OP_SUCCESS) {
+                AWS_LOGF_ERROR(
+                    AWS_LS_HTTP_STREAM,
+                    "id=%p: Incoming content-length header has invalid value.",
```
Comment (justinboswell) [inline source/decode.c]: Do you want to log what the invalid value was?

File: source/decode.c (hunk 1 above)
Comment (graebm) [reply source/decode.c]: basis 10
I thought this was your interview question

File: source/decode.c (hunk 2 above)
Comment (graebm) [reply source/decode.c]: They come in a comma separated list, and apparently empty entries are ok

File: source/decode.c (hunk 4 above)
Comment (graebm) [reply source/decode.c]: Youuuuuu ARE RIGHT!
thanks
fixin it

File: source/decode.c (hunk 3 above)
Comment (graebm) [reply source/decode.c]: I was on the fence about whether to log invalid stuff. Like, if a header is invalid, do I log the bad string? Is this any kind of security risk? Or might I spit out 1000 lines of binary data?

thoughts?

Comment (graebm) [reply source/decode.c]: logging it.
It looks lovely btw

File: source/decode.c (hunk 5 above)
Comment (graebm) [reply source/decode.c]: did

File: source/decode.c (hunk 6 above)
Comment (graebm) [reply source/decode.c]: did

File: source/decode.c (hunk 1 above)
Comment (graebm) [reply source/decode.c]: renamed function so this isn't confusing

File: source/decode.c (hunk 2 above)
Comment (justinboswell) [reply source/decode.c]: I just mean linguistically :)

**Pull request merged**